### PR TITLE
User Interface

### PR DESCRIPTION
**Pull request by Kugelschieber**

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -1,18 +1,55 @@
...
+            const time = new Date();
+            time.setSeconds(time.getSeconds()+60);
+
+            return {
+                level: 5,
+                time,
+                slots: ["99", "E7", "AD", ""],
+                codes: [
+                    {
+                        code: ["", "E7", "AD", "BD"],
```
Comment (Ablu) [inline src/App.vue:34]: What does it do when the first is empty? I though it does not matter when the sequence is started? Does it have to align with the buffer?

File: src/components/Matrix.vue [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<template>
+    <div class="matrix">
+        <h2>Code-Matrix</h2>
+        <div class="matrix-row" v-for="i in size" :key="i">
+            <div class="matrix-column" v-for="j in size" :key="j">
+                {{matrix[(i-1)*5+(j-1)]}}
```
Comment (Ablu) [inline src/components/Matrix.vue]: not adding spaces between operators?

File: src/components/Matrix.vue [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    </div>
+</template>
+
+<script lang="ts">
+    import {defineComponent} from "vue";
+    
+    export default defineComponent({
+        props: {
+            size: {type: Number, required: true},
+            matrix: {type: Array, required: true}
```
Comment (Ablu) [inline src/components/Matrix.vue]: I usually add a trailing , here for keeping the diffs cleaner

File: src/components/Timer.vue [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            time: {type: Object, required: true}
+        },
+        setup(props) {
+            const countdown = ref("60.00");
+            const progress = ref(100);
+
+            setInterval(() => {
+                const remaining = (props.time.getTime()-new Date().getTime())/1000;
+
+                if(remaining > 0) {
```
Comment (Ablu) [inline src/components/Timer.vue]: space before `(`

File: src/components/Timer.vue [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+<script lang="ts">
+    import {defineComponent, ref} from "vue";
+    
+    export default defineComponent({
+        props: {
+            time: {type: Object, required: true}
+        },
+        setup(props) {
+            const countdown = ref("60.00");
+            const progress = ref(100);
```
Comment (Ablu) [inline src/components/Timer.vue]: Any reason do you do this with refs rather than simply adding this as data property?

**Review by Ablu — COMMENTED** [review]
mostly minor stuff, but I am a bit irritated due to the `ref()` use. I would find a simple data property a lot easier? Or am I missing something here?

File: src/App.vue (hunk 1 above)
Comment (Kugelschieber) [reply src/App.vue:34]: That's used as an offset.

File: src/components/Matrix.vue (hunk 2 above)
Comment (Kugelschieber) [reply src/components/Matrix.vue]: nö :D

File: src/components/Timer.vue (hunk 5 above)
Comment (Kugelschieber) [reply src/components/Timer.vue]: ref returns an observable object and is used for two-way data binding to update the view.

Comment (Ablu) [reply src/components/Timer.vue]: Well... but you are not doing two-way data binding, right?

Comment (Kugelschieber) [reply src/components/Timer.vue]: We do, it's rendered in the template above and won't update without it.

Comment (Ablu) [reply src/components/Timer.vue]: I meant it like this: https://jsfiddle.net/api/post/library/pure/#&togetherjs=c6tacD1jpt. The timer will be likely updated from the GameState and not from the UI, right?

Comment (Kugelschieber) [reply src/components/Timer.vue]: 404... we can calculate the time the timer should end instead of updating the timer through properties all the time.

Comment (Ablu) [reply src/components/Timer.vue]: hm... does this work: https://jsfiddle.net/ofcb9spq/? Anyway, https://paste.centos.org/view/210c0516 is probably better explaining it. That way you only have a single value you need to change while the others update declaratively.

**Review by Ablu — APPROVED** [review]

**Pull request merged**